lamagraph-core/src/Core/MemoryManager.hs

```diff
@@ -0,0 +1,450 @@
+>>> :set -XAllowAmbiguousTypes
+>>> :set -XLambdaCase
+-}
+
+data EdgeEnd (portsNumber :: Nat) = EdgeEnd
+  { _addressOfVertex :: AddressNumber
+  , _idOfPort :: IdOfPort portsNumber
+  }
+  deriving (Generic, NFDataX, Show, Eq)
+
+$(makeLenses ''EdgeEnd)
+
+data Edge (portsNumber :: Nat) = Edge
+  { _leftEnd :: EdgeEnd portsNumber
+  , _rightEnd :: EdgeEnd portsNumber
+  }
+  deriving (Generic, NFDataX, Show, Eq)
+
+$(makeLenses ''Edge)
+
+data ActivePair (portsNumber :: Nat) = ActivePair
+  { _leftNode :: LoadedNode portsNumber
+  , _rightNode :: LoadedNode portsNumber
+  }
+  deriving (Show, Eq, Generic, NFDataX, Bundle)
+
+$(makeLenses ''ActivePair)
+
+data MemoryManager (cellsNumber :: Nat)
+  = MemoryManager
+  { _busyBitMap :: Vec cellsNumber Bool -- map Address : Bool. tell smth like "this Address is busy, so you can not to write here"
+  , _activePairs :: Vec cellsNumber Bool
+  }
+  deriving (Generic, NFDataX, Bundle)
+
+$(makeLenses ''MemoryManager)
+
+data Delta (nodesNumber :: Nat) (edgesNumber :: Nat) (portsNumber :: Nat) = Delta
+  { _newNodes :: Vec nodesNumber (Maybe (LocalNode portsNumber))
+  , _newEdges :: Vec edgesNumber (Maybe (Edge portsNumber))
+  , _activePair :: ActivePair portsNumber
+  }
+  deriving (Show, Eq, Generic, NFDataX)
+
+$(makeLenses ''Delta)
+
+{- | Cast `Index` to `Unsigned` if possible by adding non-significant zeros
+
+==== __Example__
+
+>>> indexToUnsigned (2 :: Index 4) :: Unsigned 16
+2
+-}
+indexToUnsigned ::
+  forall n m.
+  (KnownNat n, KnownNat m, 1 <= n, CLog 2 n <= m) =>
+  Index n ->
+  Unsigned m
+indexToUnsigned v = bitCoerce (resize v :: Index (2 ^ m))
+
+{- | Get address from memory manager that is not busy. Return `Nothing` if all addresses are busy
+
+==== __Example__
+
+>>> getUnusedAddress (repeat False :: Vec 65536 Bool)
+Just 0
+-}
+getUnusedAddress ::
+  (KnownNat cellsNumber, 1 <= cellsNumber, CLog 2 cellsNumber <= BitSize AddressNumber) =>
+  Vec cellsNumber Bool ->
+  Maybe ActualAddressNumber
+getUnusedAddress busyMap = address
+ where
+  indexOfUnused = elemIndex False busyMap
+  address = indexToUnsigned <$> indexOfUnused
+
+{- | Mark given `AddressNumber` as busy or not according to passed flag (`True` means busy)
+
+==== __Example__
+
+>>> markAddress (repeat False :: Vec 4 Bool) True 2
+False :> False :> True :> False :> Nil
+-}
+markAddress ::
+  (KnownNat cellsNumber) =>
+  Vec cellsNumber Bool ->
+  Bool ->
+  ActualAddressNumber ->
+  Vec cellsNumber Bool
+markAddress busyMap marker address =
+  replace address marker busyMap
+
+-- | Replace processed active pair at `Nothing` in `Vec` of active pairs
+deleteActivePair ::
+  (KnownNat cellsNumber, KnownNat portsNumber) =>
+  Vec cellsNumber Bool ->
+  ActivePair portsNumber ->
+  Vec cellsNumber Bool
+deleteActivePair oldActivePairs activePairToDelete =
+  if leftInVec `xor` rightInVec
+    then newActivePairs
+    else error "In active pairs map should be exact one address"
```

Comment:
I'm not sure that message is informative enough. Even more, I'm not sure that I realize what it means.